core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -727,40 +727,44 @@ private[master] class Master(
 
   def removeApplication(app: ApplicationInfo, state: ApplicationState.Value) {
     if (apps.contains(app)) {
-      logInfo("Removing app " + app.id)
-      apps -= app
-      idToApp -= app.id
-      endpointToApp -= app.driver
-      addressToApp -= app.driver.address
-      if (completedApps.size >= RETAINED_APPLICATIONS) {
-        val toRemove = math.max(RETAINED_APPLICATIONS / 10, 1)
-        completedApps.take(toRemove).foreach( a => {
-          appIdToUI.remove(a.id).foreach { ui => webUi.detachSparkUI(ui) }
-          applicationMetricsSystem.removeSource(a.appSource)
-        })
-        completedApps.trimStart(toRemove)
-      }
-      completedApps += app // Remember it in our history
-      waitingApps -= app
+      synchronized{
```

Comment:
It's a premature optimization -- why? 
Also I don't think this really addresses the full problem, as there are other methods modifying the same state concurrently. I'd rather not commit something incomplete.